File: apps/api/src/abf_api/routers/agent_runs.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from supabase import Client

from abf_api.deps.supabase import get_supabase
from abf_api.models.schemas import AgentRunCreate, AgentRunRead, AgentRunUpdate

router = APIRouter(prefix="/agent-runs", tags=["agent-runs"])


@router.get("", response_model=list[AgentRunRead])
def list_agent_runs(
    business_id: str | None = None,
    status: str | None = None,
    agent_type: str | None = None,
    db: Client = Depends(get_supabase),
):
    q = db.table("agent_runs").select("*").order("started_at", desc=True)
    if business_id:
        q = q.eq("business_id", business_id)
    if status:
        q = q.eq("status", status)
    if agent_type:
        q = q.eq("agent_type", agent_type)
    return q.execute().data


@router.get("/{run_id}", response_model=AgentRunRead)
def get_agent_run(run_id: str, db: Client = Depends(get_supabase)):
    res = db.table("agent_runs").select("*").eq("id", run_id).maybe_single().execute()
    if res.data is None:
        raise HTTPException(status_code=404, detail="Agent run not found")
    return res.data


@router.post("", response_model=AgentRunRead, status_code=201)
def create_agent_run(body: AgentRunCreate, db: Client = Depends(get_supabase)):
    data = body.model_dump(exclude_none=True)
    data["status"] = "queued"
    res = db.table("agent_runs").insert(data).execute()
    return res.data[0]


@router.patch("/{run_id}", response_model=AgentRunRead)
def update_agent_run(
    run_id: str,
    body: AgentRunUpdate,
    db: Client = Depends(get_supabase),
):
    data = body.model_dump(exclude_none=True)
    if not data:
        raise HTTPException(status_code=400, detail="No fields to update")
    if data.get("status") in ("completed", "failed"):
        data["completed_at"] = datetime.now(timezone.utc).isoformat()
    res = db.table("agent_runs").update(data).eq("id", run_id).execute()
    if not res.data:
        raise HTTPException(status_code=404, detail="Agent run not found")
    return res.data[0]

File: packages/db/src/queries/agent-runs.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type {
  Database,
  AgentRunInsert,
  AgentRunUpdate,
  AgentRunStatus,
  AgentType,
  Json,
} from "../types/database";

type Client = SupabaseClient<Database>;

export function listAgentRuns(
  client: Client,
  filters?: { businessId?: string; status?: AgentRunStatus; agentType?: AgentType }
) {
  let query = client
    .from("agent_runs")
    .select("*, businesses(name), tasks(title)")
    .order("started_at", { ascending: false });

  if (filters?.businessId) {
    query = query.eq("business_id", filters.businessId);
  }
  if (filters?.status) {
    query = query.eq("status", filters.status);
  }
  if (filters?.agentType) {
    query = query.eq("agent_type", filters.agentType);
  }

  return query;
}

export function getAgentRunById(client: Client, id: string) {
  return client
    .from("agent_runs")
    .select("*, businesses(name), tasks(title)")
    .eq("id", id)
    .single();
}

export function createAgentRun(client: Client, data: AgentRunInsert) {
  return client.from("agent_runs").insert(data).select().single();
}

export function updateAgentRun(
  client: Client,
  id: string,
  data: AgentRunUpdate
) {
  return client.from("agent_runs").update(data).eq("id", id).select().single();
}

export function completeAgentRun(
  client: Client,
  id: string,
  output: Json,
  stats: { duration_ms: number; tokens_used: number; cost_cents: number }
) {
  const update: AgentRunUpdate = {
    status: "completed" as const,
    output_payload: output,
    completed_at: new Date().toISOString(),
    ...stats,
  };
  return client.from("agent_runs").update(update).eq("id", id).select().single();
}

export function failAgentRun(
  client: Client,
  id: string,
  errorMessage: string,
  stats: { duration_ms: number; tokens_used: number; cost_cents: number }
) {
  const update: AgentRunUpdate = {
    status: "failed" as const,
    error_message: errorMessage,
    completed_at: new Date().toISOString(),
    ...stats,
  };
  return client.from("agent_runs").update(update).eq("id", id).select().single();
}
